design-documents/service-isolation/detailed-architecture.md

```diff
@@ -0,0 +1,87 @@
+# Detailed architecture of multi-service Magento
+
+## Overview
+
+The goal of this document is to gather requirements, understand data flow and propose architecture for multi-service Magento application.
+
+Please make sure to get familiar with [high-level service isolation vision](../service-isolation.md) before proceeding.
+
+![Gateway](img/GraphQL%20BFF.png)
+
+The client (e.g. browser) is interacting with the system using GraphQL. The only component of the system exposed publicly is Storefront Gateway, which is responsible for routing, full response caching and perimeter-level authorization.
+
+Storefront Gateway routes incoming queries between PWA, GraphQL Engine and Storefront Authentication Gateway.
+
+PWA's main responsibility is to [render React application on the server-side](../frontend/server-side-rendering.md). Resulting HTML page with React application becomes cacheable on the server.
+
+GraphQL server parses GraphQL query, enriches the request (explained later) and proxies calls to the Storefront services. It also manages Customer context JWT. GraphQL schema is defined here and is extensible, query resolution is based on generic resolvers and configuration for field mapping to Storefront APIs.
+
```

Comment:
Do Magento connects to 3rd party APIs? if that is the case, we need to think about that? either here or in additional proposal

Reply:
It probably does for online payments and shipment. @joni-jones could you please provide any information on this?

Reply:
Magento is not a connector to 3rd party systems itself, at least right now. But the application has multiple integrations with different 3rd party solutions like shipping carriers, payment gateways, fraud systems, etc. In this case, each service will communicate with a 3rd party system directly as only such service has knowledge of how the 3rd API looks and how to communicate with it. Also, in most cases, a separate service will expose all needed API to API gateway to avoid direct communication between API Gateway and 3rd party integration.

Reply:
There are multiple patterns here, one as suggested by @joni-jones which is okay but it does spreads 3rd party integration code to multiple individual services, each handling them in different ways possibly. 

In case of observably and security, some organization routes external calls to outgoing gateway; but I believe if we use sidecar pattern / proxy like Envoy, this might solve the security and telemetry issue without a gateway.

Reply:
Added statement about Observability